sorting/external_sort.cpp

```diff
@@ -0,0 +1,187 @@
+#include <string>
+
+using namespace std;
+
+// buff data which fit in memory to sort
+constexpr auto BUFF_SIZE = 100000;
+
+struct ChunkFile {
+    ifstream file;
+    int currentVal;
+    string line;
+};
+
+/**
+   Prepare 1M rows data (int) data being sorted and write to text file.
+ */
+void mockData() {
+    cout << "Preparing data" << endl;
+
+    // change to absolutely path if u want
+    ofstream data_file("data.txt", ofstream::trunc);  // overwite
+
+    // use to get a random number
+    mt19937 generator(time(0));
+    uniform_int_distribution<int> distribution(0, 999999);
+
+    for (int i = 0; i < 1000000; i++) {
+        // do random number
+        int randomNumber = distribution(generator);
+        data_file << randomNumber << "\n";
+    }
+    data_file.close();
+
+    cout << "End preparing data" << endl;
+}
+
+void externalSort(char* data_file) {
+    ifstream input_file(data_file);
+
+    string line;
+    int line_number = 0;
+    vector<int> chunk;
+    vector<string> chunk_file_names;
+
+    cout << "Do sort chunks" << endl;
+
+    // read chunk and sort each chunk
+    while (getline(input_file, line)) {
+        line_number++;
+        // convert line to number and add to chunk
+        chunk.push_back(stoi(line));
+        if (line_number == BUFF_SIZE) {
+            // if have enough buffer data then do sort the chunk
+            // note: i am using a built-in <algorithm> sort function,
+            // you can use any other in-memory sort algorithms such as bubble
+            // sort or quick sort
+            sort(chunk.begin(), chunk.end());  // asc
+
+            // write sort result to temporary file. use index as file name
+            string temp_file_name = to_string(chunk_file_names.size());
+            ofstream data_file(temp_file_name, ofstream::trunc);
+            for (int val : chunk) {
+                data_file << val << "\n";
+            }
+            data_file.close();
+            chunk_file_names.push_back(temp_file_name);
+
+            chunk.clear();
+            line_number = 0;
+        }
+    }
+    // sort the last chunk
+    if (line_number != 0) {
+        sort(chunk.begin(), chunk.end());
+
+        // write sort result to temporary file. use index as file name
+        string temp_file_name = to_string(chunk_file_names.size());
+        ofstream data_file(temp_file_name, ofstream::trunc);
+        for (int val : chunk) {
+            data_file << val << "\n";
+        }
+        data_file.close();
+        chunk_file_names.push_back(temp_file_name);
+    }
+
+    cout << "End sort chunks" << endl;
+    cout << "Merge sorted chunks using k-way merging" << endl;
+
+    // merge all sorted chunks to the final result, using K-Way merging
+    vector<ChunkFile*> chunk_files;
+
+    // change to absolutely path if u want
+    ofstream result_file("sort_result.txt", ofstream::trunc);
+
+    // open chunk files and read the first value
+    for (string f_name : chunk_file_names) {
```

Comment:
```suggestion
    for (string filename : chunk_file_names) {
```
there isnt a limit to length of variable names please always make sure to use descriptive names